tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -31,6 +31,163 @@ bool ProgramInfo::checkStructuralEquality(uint32_t V, uint32_t U) {
   return false;
 }
 
+bool ProgramInfo::link() {
+
+  // For every global symbol in all the global symbols that we have found
+  // go through and apply rules for whether they are functions or variables.
+
+  for (const auto &S : GlobalSymbols) {
+    // First, extract out the function symbols from S. 
+    std::set<GlobalFunctionSymbol*> funcs;
+    std::set<GlobalVariableSymbol*> vars;
+    for (const auto &U : S.second)
+      if (GlobalFunctionSymbol *K = dyn_cast<GlobalFunctionSymbol>(U))
+        funcs.insert(K);
+      else if (GlobalVariableSymbol *V = dyn_cast<GlobalVariableSymbol>(U))
+        vars.insert(V);
+
+    // Then, pairwise-iterate over each of the function symbols found for this 
+    // symbol. What we want to do is for a sequence of constraint variables on
+    // F1,F2, set the constraint variables for F1(v0,vi,vN) and F2(v0,vj,vN) to
+    // be equal to each other, i.e. to enter a series of constraints of the form
+    // vi == vj.
+    //
```

Comment:
Because this is a unification based approach, do you need to iterate pairwise overall function definitions?    Given N copies of f (f1, ... fn), doesn't it suffice to set the constraints for f1 = f2, f2 = f3 ... and so on?